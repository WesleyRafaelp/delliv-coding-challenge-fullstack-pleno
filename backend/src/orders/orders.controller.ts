import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  UseGuards,
} from '@nestjs/common';
import { OrdersService } from './orders.service';
import { CreateOrderDto } from './dto/create-order.dto';
import { UpdateOrderDto } from './dto/update-order.dto';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { JwtAuthGuard } from 'src/auth/jwt-auth.guard';

@ApiTags('orders')
@Controller('orders')
export class OrdersController {
  constructor(private readonly ordersService: OrdersService) {}

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Post()
  async create(@Body() createOrderDto: CreateOrderDto) {
    return await this.ordersService.create(createOrderDto);
  }

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Get()
  findAll() {
    return this.ordersService.findAll();
  }

  // @Get(':id')
  // findOne(@Param('id') id: string) {
  //   return this.ordersService.findOne(+id);
  // }

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Patch(':id')
  update(@Param('id') id: string, @Body() updateOrderDto: UpdateOrderDto) {
    return this.ordersService.update(+id, updateOrderDto);
  }

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.ordersService.remove(+id);
  }
}
