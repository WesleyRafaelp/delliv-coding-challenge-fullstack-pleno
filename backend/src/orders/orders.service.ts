import { Injectable } from '@nestjs/common';
import { CreateOrderDto } from './dto/create-order.dto';
import { UpdateOrderDto } from './dto/update-order.dto';
import { PrismaService } from '../prisma.service';

@Injectable()
export class OrdersService {
  constructor(private prisma: PrismaService) {}

  async create(data: CreateOrderDto) {
    const { client, deliveryAddress } = data;

    return await this.prisma.order.create({
      data: {
        client,
        deliveryAddress,
        status: 'PENDENTE',
      },
    });
  }

  findAll() {
    return this.prisma.order.findMany();
  }

  update(id: number, updateOrderDto: UpdateOrderDto) {
    return this.prisma.order.update({
      where: { id },
      data: { status: updateOrderDto.status },
    });
  }

  remove(id: number) {
    return this.prisma.order.delete({ where: { id } });
  }
}
