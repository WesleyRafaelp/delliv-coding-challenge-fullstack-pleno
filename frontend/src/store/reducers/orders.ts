import { orderService } from "@/services/order/order-service";
import { Action, Dispatch, ThunkAction, createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { RootState } from "..";
import { stat } from "fs";

export interface Order {
    id: number;
    client: string;
    deliveryAddress: string;
    status: string;
}

interface OrderState {
    orders: Order[];
    loading: boolean;
    error: string | null;
}

const initialState: OrderState = {
    orders: [],
    loading: false,
    error: null,
};

export const getOrders = createAsyncThunk('order/getOrders', orderService.getListOrder);

export const newOrder = createAsyncThunk('order/create', async (order: Partial<Order>) => {
    const res = await orderService.create(order)
    return res
})
export const updateOrder = createAsyncThunk('order/update', async (order: Partial<Order>) => {
    const res = await orderService.update(order)
    return res
})

export const removeOrder = createAsyncThunk('order/remove', async (id: any) => {
    const res = await orderService.delete(id)
    return res
})

export const ordersSlice = createSlice({
    name: "orders",
    initialState,
    reducers: {},
    extraReducers: builder => {
        builder
            .addCase(
                getOrders.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: payload,
                        loading: false,
                        error: null,
                    };
                },
            )
            .addCase(
                getOrders.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                getOrders.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
            .addCase(
                newOrder.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: [...state.orders, payload],
                        error: null,
                        loading: false,
                    };
                },
            )
            .addCase(
                newOrder.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                newOrder.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
            .addCase(
                updateOrder.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: state.orders.map(order =>
                            order.id === payload.id ? { ...order, ...payload } : order
                        ),
                        error: null,
                        loading: false,
                    };
                },
            )
            .addCase(
                updateOrder.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                updateOrder.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
            .addCase(
                removeOrder.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: state.orders.filter(order => order.id !== payload.id),
                        loading: false,
                        error: null,
                    };
                },
            )
            .addCase(
                removeOrder.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                removeOrder.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
    }
})

export default ordersSlice.reducer;
export const orderSelector = (state: RootState) => state.orderReducer;