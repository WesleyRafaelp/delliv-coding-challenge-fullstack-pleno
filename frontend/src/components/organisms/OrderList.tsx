import React, { useEffect, useState } from 'react';
import OrderStatusUpdater from '../molecules/OrderStatusUpdater';
import DeleteButton from '../atoms/DeleteButton';
import ConfirmationPopup from '../molecules/ConfirmationPopup';
import { getOrders, orderSelector, removeOrder, updateOrder } from '@/store/reducers/orders';
import { useAppDispatch, useAppSelector } from '@/hooks/hooks';

const OrderList: React.FC = () => {
  const [selectedOrderId, setSelectedOrderId] = useState<number | null>(null);

  const dispatch = useAppDispatch()
  const {orders, loading, error}  = useAppSelector(orderSelector);

  

  useEffect(() => {
    dispatch(getOrders())
  }, [dispatch])

  const handleDeleteClick = (orderId: number) => {
    setSelectedOrderId(orderId);
  };

  const handleCancelDelete = () => {
    setSelectedOrderId(null);
  };

  const handleConfirmDelete = () => {
    dispatch(removeOrder(selectedOrderId));
    setSelectedOrderId(null);
  };

  const handleUpdateStatus = (orderId: number, newStatus: string) => {
    const order = {
      id:orderId,
      status: newStatus,
    }
    dispatch(updateOrder(order))
  };

  return (
    <div>
      {loading && (
        <div className="bg-blue-100 border-t-4 border-blue-500 rounded-b text-blue-900 px-4 py-3 shadow-md mb-4" role="alert">
          <div className="flex">
            <div className="py-1">
              <svg className="fill-current h-6 w-6 text-blue-500 mr-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
                <path
                  d="M2 10A8 8 0 1 0 10 2 8 8 0 0 0 2 10zm9.293-5.293a1 1 0 0 1 1.414 0l4 4a1 1 0 0 1-1.414 1.414L11 6.414l-3.293 3.293a1 1 0 1 1-1.414-1.414l4-4z"
                />
              </svg>
            </div>
            <div>
              <p className="font-bold">Loading...</p>
            </div>
          </div>
        </div>
      )}
      {error && (
        <div className="bg-red-100 border-t-4 border-red-500 rounded-b text-red-900 px-4 py-3 shadow-md mb-4" role="alert">
          <div className="flex">
            <div className="py-1">
              <svg className="fill-current h-6 w-6 text-red-500 mr-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
                <path
                  d="M10 2C5.82 2 2 5.82 2 10s3.82 8 8 8 8-3.82 8-8-3.82-8-8-8zm1 12h-2v-2h2v2zm0-4h-2V6h2v4z"
                />
              </svg>
            </div>
            <div>
              <p className="font-bold">Error: {error}</p>
            </div>
          </div>
        </div>
      )}
      { orders && orders.map((order:any) => (
        <div key={order.id} className="bg-white border border-gray-300 rounded p-4 mb-4 shadow-md">
          <div className="mb-2">
            <strong className="text-lg font-semibold text-gray-800">Cliente:</strong> {order.client}
          </div>
          <div className="mb-2">
            <strong className="text-lg font-semibold text-gray-800">Endere√ßo de Entrega:</strong> {order.deliveryAddress}
          </div>
          <div className="mb-2">
            <strong className="text-lg font-semibold text-gray-800">Status:</strong> {order.status}
          </div>
          <div className="mt-4 flex justify-between space-x-2">
            <OrderStatusUpdater
              currentStatus={order.status}
              onUpdateStatus={(newStatus) => handleUpdateStatus(order.id, newStatus)}
              availableStatusOptions={['PENDENTE', 'EM_PROCESSAMENTO', 'ENVIADO', 'ENTREGUE', 'CANCELADO']}
            />
            <DeleteButton onClick={() => handleDeleteClick(order.id)} />
          </div>
        </div>
      ))}
      <ConfirmationPopup
        isOpen={selectedOrderId !== null}
        onCancel={handleCancelDelete}
        onConfirm={handleConfirmDelete}
      />
    </div>
  );
};

export default OrderList;
