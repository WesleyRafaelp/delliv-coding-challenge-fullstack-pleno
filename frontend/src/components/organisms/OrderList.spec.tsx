import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import OrderList from './OrderList';
import { useAppDispatch, useAppSelector } from '@/hooks/hooks';
import { getOrders, removeOrder, updateOrder } from '@/store/reducers/orders';

jest.mock('@/hooks/hooks');
jest.mock('@/store/reducers/orders');

describe('OrderList Component', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  it('renders OrderList component', () => {
    (useAppSelector as jest.Mock).mockReturnValue({
      orders: [
        { id: 1, client: 'John Doe', deliveryAddress: '123 Main St', status: 'PENDENTE' },
        { id: 2, client: 'Jane Doe', deliveryAddress: '456 Oak St', status: 'ENVIADO' },
      ],
      loading: false,
      error: null,
    });

    render(<OrderList />);

    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByText('Jane Doe')).toBeInTheDocument();
  });

  it('displays loading message', () => {
    (useAppSelector as jest.Mock).mockReturnValue({
      orders: [],
      loading: true,
      error: null,
    });

    render(<OrderList />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('displays error message', () => {
    const errorMessage = 'Some error message';
    (useAppSelector as jest.Mock).mockReturnValue({
      orders: [],
      loading: false,
      error: errorMessage,
    });

    render(<OrderList />);

    expect(screen.getByText(`Error: ${errorMessage}`)).toBeInTheDocument();
  });

  it('handles order status update', async () => {
    (useAppSelector as jest.Mock).mockReturnValue({
      orders: [{ id: 1, client: 'John Doe', deliveryAddress: '123 Main St', status: 'PENDENTE' }],
      loading: false,
      error: null,
    });

    render(<OrderList />);

    fireEvent.click(screen.getByText('Editar Status'));
    fireEvent.click(screen.getByText('Atualizar Status'));

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(updateOrder({ id: 1, status: 'EM_PROCESSAMENTO' }));
    });
  });

  it('handles order deletion', async () => {
    (useAppSelector as jest.Mock).mockReturnValue({
      orders: [{ id: 1, client: 'John Doe', deliveryAddress: '123 Main St', status: 'PENDENTE' }],
      loading: false,
      error: null,
    });

    render(<OrderList />);

    fireEvent.click(screen.getByText('Excluir'));

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(removeOrder(1));
    });
  });
});

describe('OrderList Component Snapshot', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  it('matches snapshot', () => {
    (useAppSelector as jest.Mock).mockReturnValue({
      orders: [
        { id: 1, client: 'John Doe', deliveryAddress: '123 Main St', status: 'PENDENTE' },
        { id: 2, client: 'Jane Doe', deliveryAddress: '456 Oak St', status: 'ENVIADO' },
      ],
      loading: false,
      error: null,
    });

    const { asFragment } = render(<OrderList />);
    expect(asFragment()).toMatchSnapshot();
  });
});