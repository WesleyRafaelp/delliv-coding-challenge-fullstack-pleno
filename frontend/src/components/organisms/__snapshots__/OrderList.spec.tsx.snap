// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList Component Snapshot matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="bg-white border border-gray-300 rounded p-4 mb-4 shadow-md"
    >
      <div
        class="mb-2"
      >
        <strong
          class="text-lg font-semibold text-gray-800"
        >
          Cliente:
        </strong>
         John Doe
      </div>
      <div
        class="mb-2"
      >
        <strong
          class="text-lg font-semibold text-gray-800"
        >
          Endereço de Entrega:
        </strong>
         123 Main St
      </div>
      <div
        class="mb-2"
      >
        <strong
          class="text-lg font-semibold text-gray-800"
        >
          Status:
        </strong>
         PENDENTE
      </div>
      <div
        class="mt-4 flex justify-between space-x-2"
      >
        <div>
          <button
            class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline-blue active:bg-blue-800"
            type="button"
          >
            Editar Status
          </button>
        </div>
        <button
          class="text-red-500 hover:text-red-700"
        >
          Excluir
        </button>
      </div>
    </div>
    <div
      class="bg-white border border-gray-300 rounded p-4 mb-4 shadow-md"
    >
      <div
        class="mb-2"
      >
        <strong
          class="text-lg font-semibold text-gray-800"
        >
          Cliente:
        </strong>
         Jane Doe
      </div>
      <div
        class="mb-2"
      >
        <strong
          class="text-lg font-semibold text-gray-800"
        >
          Endereço de Entrega:
        </strong>
         456 Oak St
      </div>
      <div
        class="mb-2"
      >
        <strong
          class="text-lg font-semibold text-gray-800"
        >
          Status:
        </strong>
         ENVIADO
      </div>
      <div
        class="mt-4 flex justify-between space-x-2"
      >
        <div>
          <button
            class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline-blue active:bg-blue-800"
            type="button"
          >
            Editar Status
          </button>
        </div>
        <button
          class="text-red-500 hover:text-red-700"
        >
          Excluir
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
